Migrate routing and navigation docs to new developer guide (#113919)
elastic/kibana: dev_docs/tutorials/endpoints.mdx

---
id: kibDevTutorialServerEndpoint
slug: /kibana-dev-docs/tutorials/registering-endpoints
title: Registering and accessing an endpoint
summary: Learn how to register a new endpoint and access it
date: 2021-10-05
tags: ['kibana', 'dev', 'architecture', 'tutorials']
---

> ## Registering an endpoint

Reviewer: Might be worth linking to the example plugin - https://demo.kibana.dev/8.0/app/routingExample?

- to extend the Kibana server with custom HTTP API.
- to execute custom logic on an incoming request or server response.
- to implement custom authentication and authorization strategy.

<DocCallOut>
  See [the server-side HTTP service API docs](https://github.com/elastic/kibana/blob/master/docs/development/core/server/kibana-plugin-core-server.httpservicesetup.md)
</DocCallOut>

**Registering a basic GET endpoint**

```ts
import { schema } from '@kbn/config-schema';
import type { CoreSetup, Plugin } from 'kibana/server';

export class MyPlugin implements Plugin {
  public setup(core: CoreSetup) {
    const router = core.http.createRouter();
    
    const validate = {
      params: schema.object({
        id: schema.string(),
      }),
    };
    
    router.get({
      path: '/api/my_plugin/{id}',
      validate
    },
    async (context, request, response) => {
      const data = await findObject(request.params.id);
      if (!data) return response.notFound();
      return response.ok({
        body: data,
        headers: {
          'content-type': 'application/json'
        }
      });
    });
  }
}
```

## Consuming the endpoint from the client-side

> The HTTP service is also offered on the client side and provides an API to communicate with the Kibana server via HTTP interface.

Reviewer 2: ```suggestion
The HTTP service provides a client side API to communicate with the Kibana server via HTTP interface.
```

The client-side `HttpService` is a preconfigured wrapper around `window.fetch` that includes some default behavior and automatically handles common errors (such as session expiration). 

> The service should only be used for access to backend endpoints registered by the same plugin. Feel free to use another HTTP client library to request 3rd party services.

Reviewer 2: Maybe we need to emphasize this more
